romancal/datamodels/container.py

```diff
@@ -0,0 +1,486 @@
+        metadata is used for grouping:
+
+        meta.observation.program
+        meta.observation.observation
+        meta.observation.visit
+        meta.observation.visit_file_group
+        meta.observation.visit_file_sequence
+        meta.observation.visit_file_activity
+        meta.observation.exposure
+        """
+        unique_exposure_parameters = [
+            "program",
+            "observation",
+            "visit",
+            "visit_file_group",
+            "visit_file_sequence",
+            "visit_file_activity",
+            "exposure",
+        ]
+
+        group_dict = OrderedDict()
+        for i, model in enumerate(self._models):
+            model = model if isinstance(model, rdm.DataModel) else rdm.open(model)
+
+            if not self._save_open:
+                model = rdm.open(model, memmap=self._memmap)
+
+            params = [
+                str(getattr(model.meta.observation, param))
+                for param in unique_exposure_parameters
+            ]
+            try:
+                group_id = "roman" + "_".join(
+                    ["".join(params[:3]), "".join(params[3:6]), params[6]]
+                )
+                model.meta["group_id"] = group_id
+            except TypeError:
+                model.meta["group_id"] = f"exposure{i + 1:04d}"
+
+            group_id = model.meta.group_id
+            if not self._save_open and not self._return_open:
+                model.close()
+                model = self._models[i]
+
+            if group_id in group_dict:
+                group_dict[group_id].append(model)
+            else:
+                group_dict[group_id] = [model]
+
+        return group_dict.values()
+
+    @property
+    def to_association(self):
+        pass
+
+    def merge_tree(self, a, b):
+        """
+        Merge elements from tree ``b`` into tree ``a``.
+        """
+
+        def recurse(a, b):
+            if isinstance(b, dict):
+                if not isinstance(a, dict):
+                    return copy.deepcopy(b)
+                for key, val in b.items():
+                    a[key] = recurse(a.get(key), val)
+                return a
+            return copy.deepcopy(b)
+
+        recurse(a, b)
+        return a
+
+    @property
+    def crds_observatory(self):
+        """
+        Get the CRDS observatory for this container.  Used when selecting
+        step/pipeline parameter files when the container is a pipeline input.
+
+        Returns
+        -------
+        str
+        """
+        return "roman"
+
+    @property
+    def get_crds_parameters(self):
+        """
+        Get parameters used by CRDS to select references for this model.
+
+        Returns
+        -------
+        dict
+        """
+        crds_header = {}
+        if len(self._models):
+            model = self._models[0]
+            model = model if isinstance(model, rdm.DataModel) else rdm.open(model)
+            crds_header |= model.get_crds_parameters()
```

Comment:
I think this will work, I'm just not sure why it is necessary when a model is guaranteed to have a `get_crds_parameters` method which returns a valid dictionary. Perhaps it's best to avoid the extra computation, just return the output of that method.